model/inception3d.py

```diff
@@ -214,140 +265,211 @@ def __init__(self, num_classes=400, spatial_squeeze=True,
+        )
 
         self.build()
 
+        self.modules_dict = nn.ModuleDict(self._modules)
 
     def replace_logits(self, num_classes):
         self._num_classes = num_classes
-        self.logits = Unit3D(in_channels=384+384+128+128, output_channels=self._num_classes,
-                             kernel_shape=[1, 1, 1],
-                             padding=0,
-                             activation_fn=None,
-                             use_batch_norm=False,
-                             use_bias=True,
-                             name='logits')
+        self.logits = Unit3D(
+            in_channels=384 + 384 + 128 + 128,
+            output_channels=self._num_classes,
+            kernel_shape=[1, 1, 1],
+            padding=0,
+            activation_fn=None,
+            use_batch_norm=False,
+            use_bias=True,
+            name="logits",
+        )
 
     def build(self):
         for k in self.end_points.keys():
             self.add_module(k, self.end_points[k])
-        
+
     def forward(self, x, pretrained=False, n_tune_layers=-1):
         if pretrained:
             assert n_tune_layers >= 0
 
-            freeze_endpoints = self.VALID_ENDPOINTS[:-n_tune_layers]
-            tune_endpoints = self.VALID_ENDPOINTS[-n_tune_layers:]
+            freeze_endpoints: List[str] = self.VALID_ENDPOINTS[:-n_tune_layers]
+            tune_endpoints: List[str] = self.VALID_ENDPOINTS[-n_tune_layers:]
         else:
-            freeze_endpoints = []
-            tune_endpoints = self.VALID_ENDPOINTS
+            freeze_endpoints: List[str] = []
+            tune_endpoints: List[str] = self.VALID_ENDPOINTS
 
         # backbone, no gradient part
         with torch.no_grad():
             for end_point in freeze_endpoints:
-                if end_point in self.end_points:
-                    x = self._modules[end_point](x) # use _modules to work with dataparallel
+                for k, v in self.end_points.items():
+                    if end_point == k:
+                        x = v(x)
+                # if end_point in self.end_points.keys():
+
+                #     x = self.modules_dict[end_point](
+                #         x
+                #     )  # use _modules to work with dataparallel
 
         # backbone, gradient part
         for end_point in tune_endpoints:
-            if end_point in self.end_points:
-                x = self._modules[end_point](x) # use _modules to work with dataparallel
+            for k, v in self.end_points.items():
+                if end_point == k:
+                    x = v(x)
+            # if end_point in self.end_points.keys():
+            #     x = self.modules_dict[end_point](
+            #         x
+            #     )  # use _modules to work with dataparallel
 
         # head
         x = self.logits(self.dropout(self.avg_pool(x)))
-        if self._spatial_squeeze:
-            logits = x.squeeze(3).squeeze(3)
+        # if self._spatial_squeeze:
```

Comment:
😮 Why did we get rid of that `if` statement?

Reply:
because it's always true

Reply:
but fixed this too